qat graph mode: initial scaffolding for prepare_qat_script

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* #39837 eager qat: refactor conv-bn to make conv with scaled weights a func
* #39840 qat graph mode: make `_forward` of per channel observers scriptable
* **#39796 qat graph mode: initial scaffolding for prepare_qat_script**

Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

vkuzo committed:
qat graph mode: initial scaffolding for prepare_qat_script

Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/quantization_type.h:
```diff
@@ -5,7 +5,7 @@ namespace jit {
 // Quantization type (dynamic quantization, static quantization).
 // Should match the Python enum in quantize_script.py
-enum class QuantType{ DYNAMIC, STATIC };
+enum class QuantType{ DYNAMIC, STATIC, QAT };
```
I was mostly trying to say that QAT might just be sharing all the passes with Static, but it's probably fine since we also share most of the static code in dynamic as well.

vkuzo replied:
yeah, hopefully most of the code can be shared.  It's mostly for readability, so that we state intent explicitly that something works for both static and qat, such as

```
if (quant_type == QuantType::STATIC || quant_type == QuantType::QAT) {
  ...
}
```

jerryzh168 replied:
that might be around the whole code base, I think we should just do the check whenever we have special path for QAT

jerryzh168 commented on test/quantization/test_quantize_script.py:
```diff
@@ -2615,3 +2617,29 @@ def forward(self, x):
+
+class TestQuantizeQATScript(QuantizationTestCase):
+
+    def test_prepare_qat(self):
```
maybe we don't need to explicitly test for this since it is identical with static

jerryzh168 replied:
I assume this is a place holder for other tests

vkuzo replied:
it might make more sense in a PR or two, adding it here mostly for easy devex.  I'm flexible.  It will call a different pass for conv-bn in the future, so not exactly the same as static.

jerryzh168 replied:
I think you can just test for the fold bn functionality, that is orthogonal to what is been tested here

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]

vkuzo committed:
Update on "qat graph mode: initial scaffolding for prepare_qat_script"


Summary:

Adds an entry point into prepare_script for QAT. No logic change yet,
just setting up the callsite.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21981644](https://our.internmc.facebook.com/intern/diff/D21981644)

[ghstack-poisoned]
